ethToElrond/v2/ethToElrondBridgeExecutor.go

```diff
@@ -0,0 +1,206 @@
+	return executor.log
+}
+
+// MyTurnAsLeader returns true if the current relayer node is the leader
+func (executor *ethToElrondBridgeExecutor) MyTurnAsLeader() bool {
+	return executor.topologyProvider.MyTurnAsLeader()
+}
+
+// GetAndStoreActionID fetches the action ID by using the stored batch and will return and store the action ID
+func (executor *ethToElrondBridgeExecutor) GetAndStoreActionID(ctx context.Context) (uint64, error) {
+	if executor.batch == nil {
+		return 0, errNilBatch
+	}
+
+	actionID, err := executor.elrondClient.GetActionIDForProposeTransfer(ctx, executor.batch)
+	if err != nil {
+		return 0, err
+	}
+
+	executor.actionID = actionID
+
+	return actionID, nil
+}
+
+// GetStoredActionID will return the stored action ID
+func (executor *ethToElrondBridgeExecutor) GetStoredActionID() uint64 {
+	return executor.actionID
+}
+
+// GetAndStoreBatchFromEthereum will fetch and store the batch from the ethereum client
+func (executor *ethToElrondBridgeExecutor) GetAndStoreBatchFromEthereum(ctx context.Context, nonce uint64) error {
+	batch, err := executor.ethereumClient.GetBatch(ctx, nonce)
+	// TODO add error filtering here
+	if err != nil {
+		return err
+	}
+
+	executor.batch = batch
+
+	return nil
+}
+
+// GetStoredBatch returns the stored batch
+func (executor *ethToElrondBridgeExecutor) GetStoredBatch() *clients.TransferBatch {
+	return executor.batch
+}
+
+// GetLastExecutedEthBatchIDFromElrond returns the last executed batch ID that is stored on the Elrond SC
+func (executor *ethToElrondBridgeExecutor) GetLastExecutedEthBatchIDFromElrond(ctx context.Context) (uint64, error) {
+	return executor.elrondClient.GetLastExecutedEthBatchID(ctx)
+}
+
+// VerifyLastDepositNonceExecutedOnEthereumBatch will check the deposit nonces from the fetched batch from Ethereum client
+func (executor *ethToElrondBridgeExecutor) VerifyLastDepositNonceExecutedOnEthereumBatch(ctx context.Context) error {
+	if executor.batch == nil {
+		return errNilBatch
+	}
+
+	lastNonce, err := executor.elrondClient.GetLastExecutedEthTxID(ctx)
+	if err != nil {
+		return err
+	}
+
+	return executor.verifyDepositNonces(lastNonce)
+}
+
+func (executor *ethToElrondBridgeExecutor) verifyDepositNonces(lastNonce uint64) error {
+	startNonce := lastNonce + 1
+	for _, dt := range executor.batch.Deposits {
+		if dt.Nonce != startNonce {
+			return fmt.Errorf("%w for deposit %s", errInvalidDepositNonce, dt.String())
+		}
+
+		startNonce++
+	}
+
+	return nil
+}
+
+// WasTransferProposedOnElrond checks if the transfer was proposed on Elrond
+func (executor *ethToElrondBridgeExecutor) WasTransferProposedOnElrond(ctx context.Context) (bool, error) {
+	if executor.batch == nil {
+		return false, errNilBatch
+	}
+
+	return executor.elrondClient.WasProposedTransfer(ctx, executor.batch)
+}
+
+// ProposeTransferOnElrond will propose the transfer on Elrond
+func (executor *ethToElrondBridgeExecutor) ProposeTransferOnElrond(ctx context.Context) error {
```

Comment:
Just a question for myself: No need to also get and use(in the future) the hash from this executor?

Reply:
as seen in v1 implementation, we do not need the hash for the transaction. It might be used in later versions as we might send it on the p2p. If we need the hash, we will return it.

Reply:
Understood